### ISSUE TITLE
Add a TextEncoder for prettier console output

### ISSUE THREAD

**Issue by akshayjshah** [uber-go/zap#59]
Add a TextEncoder for pretty, non-JSON console output. This is definitely a nice-to-have for the wider community, and it forces us to keep loggers and encoders strongly separate.

**Comment by tschottdorf** [comment]
Came here after looking at the code because `jsonLogger` threw away a lot of generality. Seems that as a first step, one should rename `jsonLogger` -> `genericLogger` and turn the output format into an option (so `NewJSON(...Option)` -> `New(...Option)`. Is there anything to be done after except for providing a different `encoder` impl?

**Comment by akshayjshah** [comment]
Hey @tschottdorf - yes, you're mostly right. There's not much to do here except exporting the `encoder` interface, making the `jsonLogger` type a generic `logger`, and allowing the user to pass the encoder as an option.

I'm in the middle of cleaning up some of the logger interfaces (so there's less copypasta for user-defined encoders and loggers). I'm planning to add a console-friendly text encoder as soon as that's done. I'd love some help with that, especially since you've got some actual user feedback from the cockroachDB project.

How about I ping you on this issue once we're ready to start the plain-text encoder? If you have time, take a crack at it; if not, I'll implement one and ping you on the PR for feedback.

**Comment by tamird** [comment]
@akshayjshah I'm going to take this over for @tschottdorf on the CockroachDB side.

I'd like to start prototyping moving us over to zap - do you have an estimate of when this might be ready to go? Looks like this issue is a little over a month old, so depending on the expected ETA I may just fork and hack up a text encoder for the time being.

**Comment by akshayjshah** [comment]
Cool - I've been following cockroachdb for a while, so this is exciting! ETA for this is somewhere between a week and a month - it's part of the 1.0 release, which I'd really like to finish up soon.

**Comment by magicalbanana** [comment]
If I add this will Über reach out to me for a job? Haha. 💃 🐝

**Comment by blampe** [comment]
@magicalbanana maybe if we merge it :)

For a quick update here - @akshayjshah had some big API changes planned for #33 in anticipation of the 1.0.0 milestone. The plan was to tackle those before addressing this feature since it would help simplify the implementation. If you're interested in helping out here please feel free to reach out to @akshayjshah directly (you can find his email through his GH profile) and he can provide more detail on what he was envisioning.

**Comment by tamird** [comment]
@blampe: @akshayjshah's email is not publicly visible on his GH profile. Perhaps one of you could file an issue detailing your ideas?

**Comment by akshayjshah** [comment]
Hey @magicalbanana - you can ping me at shah@uber.com. API changes should be mostly complete in the next few days, so this issue should be unblocked soon.

**Comment by akabos** [comment]
Hey,

Please see https://github.com/akabos/zaptextenc

I created text encoder for own needs but decided to release it to the public. It lacks implementations for `AddObject` and `AddMarshaler` mostly because I don't need them. Other then that it should be in usable condition and is generally faster then native JSON encoder.

I'm still very new to Go, so any feedback is appreciated.

**Comment by sectioneight** [comment]
@akabos I just merged "native" text output into zap. Will you please give it a try and let us know what you think of the API?
